### PR TITLE
[Bug #1326596]Show track title in window title bar

### PR DESCRIPTION
**Pull request by jercaianu**
Related to pull request #530 
Reworked the pull request. Now everything looks cleaner.

https://bugs.launchpad.net/mixxx/+bug/1326596

### CODE REVIEW

File: src/mixxx.h [hunk 1]
```diff
@@ -157,6 +160,7 @@ class MixxxMainWindow : public QMainWindow {
     virtual void closeEvent(QCloseEvent *event);
     virtual bool event(QEvent* e);
 
+
```
Comment (kain88-de) [inline src/mixxx.h]: please revert this line

File: src/mixxx.h [hunk 2]
```diff
@@ -150,6 +152,7 @@ class MixxxMainWindow : public QMainWindow {
     // used to uncheck the menu when the dialog of develeoper tools is closed
     void developerToolsDlgClosed(int r);
     void closeDeveloperToolsDlgChecked(int r);
+    void currentPlayingTrackChanged(QString);
```
Comment (kain88-de) [inline src/mixxx.h]: you still don't use this function so please delete this line.

**Comment by kain88-de** [comment]
@jercaianu There wasn't any need to open another PR. Next time you can just add more commits to your branch.

**Comment by jercaianu** [comment]
Thank you for your advice. I messed up git and that's why I opened another pull request. I'll be more careful next time :)

File: src/mixxx.cpp [hunk 3]
```diff
@@ -1459,6 +1463,11 @@ void MixxxMainWindow::initActions()
     }
 }
 
+void MixxxMainWindow::changeWindowTitle(QString title)
+{
+    this->setWindowTitle(title);
```
Comment (daschuer) [inline src/mixxx.cpp]: "this->" is redundant. You can probably call setWindowTitle() directly.

Comment (daschuer) [inline src/mixxx.cpp]: Ah no, we need it. Please append the original title:  " - Mixxx VERSION" 
This seams to be a common practice in Gedit and Firefox.

**Comment by daschuer** [comment]
Hi @jercaianu, 

Thank you fro your work. 

If you have fun to go an extra mile. It would be nice to refactor the signal handling a bit and get around the m_pLibrary->getSetlogFeature() call. We will probably need it for bug https://bugs.launchpad.net/mixxx/+bug/1326653

Instead of hocking into SetlogFeature::slotPlayingDeckChanged() you may move the 
emit(currentPlayingTrackChanged(currentPlayingTrack->getTitle())); call into PlayerInfo
This is a singleton class that can be accessed from everywhere. 
The history can than hock to the same new currentPlayingTrackChanged(TrackPointer pTrack) signal. The signal should emit the track pointer as data. 

```
    connect(&PlayerInfo::instance(), SIGNAL(currentPlayingTrackChanged(TrackPointer)),
            this, SLOT(slotCurrentPlayingTrackChanged(TrackPointer)));
```

**Comment by jercaianu** [comment]
Thank you for your feedback! I got rid of the getSetlogFeature call and used PlayerInfo just as you suggested. I also added at the end of the title the current Mixxx version and saved it as m_MixxxVersion. I also display the artist of the track and in case the artist field is empty, I only display the title of the track.

File: src/mixxx.cpp [hunk 4]
```diff
@@ -1459,6 +1464,25 @@ void MixxxMainWindow::initActions()
...
 
+void MixxxMainWindow::slotChangeWindowTitle(TrackPointer pTrack)
+{
+    QString title;
+    if (pTrack) {
+        QString trackTitle = pTrack->getTitle();
+        QString trackArtist = pTrack->getArtist();
+
+        if (trackArtist.size() != 0) {
+            title = trackTitle + " by " + trackArtist;
```
Comment (daschuer) [inline src/mixxx.cpp]: add tr() exclude the " " and add a context comment 
//: comment

**Comment by daschuer** [comment]
Nice! Except the translation issue it LGTM 
@kain88-de: ready for merge? 

@jercaianu: if you change the connection to setlog feature to the new signal, is emit(currentPlayingDeckChanged(maxDeck)) still required?

**Comment by jercaianu** [comment]
Hopefully i didn't forget anything. Hope it's ok now

File: src/mixxx.h [hunk 5]
```diff
@@ -141,6 +142,8 @@ class MixxxMainWindow : public QMainWindow {
     // can alert the user if a mic is not configured.
     void slotTalkoverChanged(int);
 
+    void slotChangeWindowTitle(TrackPointer pTrack);
```
Comment (ywwg) [inline src/mixxx.h]: slotUpdateWindowTitle

File: src/mixxx.cpp [hunk 6]
```diff
@@ -1510,6 +1515,27 @@ void MixxxMainWindow::initActions()
...
+{
+    QString title;
+    if (pTrack) {
+        QString trackTitle = pTrack->getTitle();
+        QString trackArtist = pTrack->getArtist();
+
+        //Displaying title as "Track Title" by "Artist" - "Current Version"
+        //Or "Track Title - "Current Version" in case the artist field is empty
+        if (trackArtist.size() != 0) {
+            title = trackTitle + tr(" by ") + trackArtist;
```
Comment (ywwg) [inline src/mixxx.cpp]: tr("%1 by %2").args... will allow for better translation.

**Comment by ywwg** [comment]
yeah I agree that we can probably get rid of the current-playing-deck signal since that's kind of useless.

My only worry is about lifetime issues of the trackpointer.  Is there a way the pointer could be invalid by the time we try to access it?  probably not but I'm trying to think it through.

File: src/mixxx.cpp (hunk 6 above)
Comment (ywwg) [inline src/mixxx.cpp]: although personally I'd just prefer "Artist - Title" and not the word "by".

Comment (jercaianu) [inline src/mixxx.cpp]: I just saw how Banshee displayed the title, so I thought is was a good idea.

Comment (esbrandt) [inline src/mixxx.cpp]: > although personally I'd just prefer "Artist - Title" and not the word "by".

+1

**Comment by kain88-de** [comment]
> My only worry is about lifetime issues of the trackpointer. Is there a way the pointer could be invalid by the time we try to access it? probably not but I'm trying to think it through.

it shouldn't. We are using it as a shared pointer for it. As I understand the signal system Qt will keep a reference to the pointer until the signal is delivered. If we manage to access an invalid shared pointer we did screw up somewhere badly.

File: src/mixxx.cpp [hunk 7]
```diff
@@ -1510,6 +1515,27 @@ void MixxxMainWindow::initActions()
...
+        QString trackArtist = pTrack->getArtist();
+
+        //Displaying title as "Track Title" by "Artist" - "Current Version"
+        //Or "Track Title - "Current Version" in case the artist field is empty
+        if (trackArtist.size() != 0) {
+            title = trackArtist + " - " + trackTitle;
+        } else {
+            title = trackTitle;
+        }
+        title = title + " - " + m_MixxxVersion;
```
Comment (kain88-de) [inline src/mixxx.cpp]: I think you should use another sign then `-` to seperate the mixxx version. It is a little bit confusing to see the minus twice. Maybe write the string like this `artist: song  -  Mixxx`. (note the two spaces around -).

File: src/mixxx.cpp [hunk 8]
```diff
@@ -1510,6 +1515,27 @@ void MixxxMainWindow::initActions()
...
+    } else {
+        title = m_MixxxVersion;
```
Comment (kain88-de) [inline src/mixxx.cpp]: if you initialize the variable with this you can get rid of the else clause.

**Comment by kain88-de** [comment]
lgtm besides of the minor comments.

File: src/mixxx.cpp (hunk 7 above)
Comment (daschuer) [inline src/mixxx.cpp]: This is probably the Reason Banshee uses "by". This also fixes the other issue that you see only the first snipped in the task bar, the track title is more significant than the artist.

Comment (jercaianu) [inline src/mixxx.cpp]: So, how would it be more appropriate, because I see many different opinions :)

Comment (kain88-de) [inline src/mixxx.cpp]: the banshee option seems good.

**Comment by rryan** [comment]
@jercaianu the tests failed due to a problem unrelated to your code (tests failed in master as well). If you merge with master then it should be fixed.

File: src/mixxx.cpp (hunk 7 above)
Comment (rryan) [inline src/mixxx.cpp]: I think we should avoid needing a translation for this feature. What about:

"$artist - $title | Mixxx"

Comment (daschuer) [inline src/mixxx.cpp]: Just noticed the GitHub title bar above and it looks somehow smooth.  
"$artist · $title - Mixxx"

On the other hand "$title by $artist - Mixxx" already works for me and has the advantage that we see the $titel... in a truncated taskbar button. 
This was @jercaianu first solution, he did already a cycle and I am afraid he likes to kill us for and extra cycles driven by such taste discussion.

**Comment by rryan** [comment]
Thanks for the feature @jercaianu! 

We may need to remove the 'by' for i18n reasons but otherwise LGTM.

**Comment by ywwg** [comment]
I like the | option.  Or, we could use different dashes:  Artist - Title — Mixxx

**Comment by ywwg** [comment]
the suggestion in the bug of a bullet also works

**Comment by rryan** [comment]
I switched us to using ```TrackInfoObject::getInfo``` which is already in charge of formatting an artist, title pair across Mixxx. Currently that uses a comma to separate -- we can change that and it will change everywhere Mixxx does this formatting.

**Comment by rryan** [comment]
Except live broadcasting which does it's own thing ;).

**Pull request merged**